File: tour-booking-frontend/src/services/reviewService.js
import api from './api';
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080/api';

const reviewService = {
    // Create review (authenticated)
    createReview: async (tourId, userId, rating, comment, bookingId = null) => {
        try {
            let url = `/reviews?tourId=${tourId}&userId=${userId}&rating=${rating}&comment=${encodeURIComponent(comment)}`;
            if (bookingId) {
                url += `&bookingId=${bookingId}`;
            }
            const response = await api.post(url);
            return response.data;
        } catch (error) {
            console.error('Error creating review:', error);
            throw error;
        }
    },

    // Get all reviews
    getAllReviews: async () => {
        const response = await axios.get(`${API_BASE_URL}/reviews`);
        return response.data;
    },

    // Get review by ID
    getReviewById: async (id) => {
        const response = await axios.get(`${API_BASE_URL}/reviews/${id}`);
        return response.data;
    },

    // Get reviews for a tour (public)
    getTourReviews: async (tourId) => {
        try {
            const response = await axios.get(`${API_BASE_URL}/reviews/tour/${tourId}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching tour reviews:', error);
            throw error;
        }
    },

    // Get reviews by user
    getUserReviews: async (userId) => {
        const response = await api.get(`/reviews/user/${userId}`);
        return response.data;
    },

    // Get tour rating stats
    getTourRatingStats: async (tourId) => {
        try {
            const response = await axios.get(`${API_BASE_URL}/reviews/stats/tour/${tourId}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching rating stats:', error);
            return { averageRating: 0, totalReviews: 0 };
        }
    },

    // Update review
    updateReview: async (reviewId, userId, rating, comment) => {
        const response = await api.put(`/reviews/${reviewId}?userId=${userId}&rating=${rating}&comment=${encodeURIComponent(comment)}`);
        return response.data;
    },

    // Delete review
    deleteReview: async (reviewId, userId, isAdmin = false) => {
        const response = await api.delete(`/reviews/${reviewId}?userId=${userId}&isAdmin=${isAdmin}`);
        return response.data;
    }
};

export default reviewService;

File: tour-service/tour-service/src/main/java/com/tourbooking/tourservice/service/ReviewService.java
package com.tourbooking.tourservice.service;

import com.tourbooking.tourservice.exception.BadRequestException;
import com.tourbooking.tourservice.exception.ResourceNotFoundException;
import com.tourbooking.tourservice.model.Booking;
import com.tourbooking.tourservice.model.Review;
import com.tourbooking.tourservice.model.Tour;
import com.tourbooking.tourservice.model.User;
import com.tourbooking.tourservice.repository.BookingRepository;
import com.tourbooking.tourservice.repository.ReviewRepository;
import com.tourbooking.tourservice.repository.TourRepository;
import com.tourbooking.tourservice.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
public class ReviewService {

    @Autowired
    private ReviewRepository reviewRepository;

    @Autowired
    private TourRepository tourRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private BookingRepository bookingRepository;

    // Create review
    @Transactional
    public Review createReview(Long tourId, Long userId, Integer rating, String comment, Long bookingId) {
        // Check tour exists
        Tour tour = tourRepository.findById(tourId)
                .orElseThrow(() -> new ResourceNotFoundException("Tour not found with id: " + tourId));

        // Check user exists
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found with id: " + userId));

        // Check if user already reviewed this tour
        if (reviewRepository.existsByTourIdAndUserId(tourId, userId)) {
            throw new BadRequestException("You have already reviewed this tour!");
        }

        // Optional: Check if user has completed booking for this tour
        if (bookingId != null) {
            Booking booking = bookingRepository.findById(bookingId)
                    .orElseThrow(() -> new ResourceNotFoundException("Booking not found with id: " + bookingId));

            if (!booking.getUser().getId().equals(userId)) {
                throw new BadRequestException("This booking does not belong to you!");
            }

            if (!booking.getTour().getId().equals(tourId)) {
                throw new BadRequestException("This booking is not for this tour!");
            }
        }

        // Create review
        Review review = new Review();
        review.setTour(tour);
        review.setUser(user);
        review.setRating(rating);
        review.setComment(comment);

        if (bookingId != null) {
            Booking booking = bookingRepository.findById(bookingId).orElse(null);
            review.setBooking(booking);
        }

        return reviewRepository.save(review);
    }

    // Get all reviews
    public List<Review> getAllReviews() {
        return reviewRepository.findAll();
    }

    // Get review by ID
    public Optional<Review> getReviewById(Long id) {
        return reviewRepository.findById(id);
    }

    // Get reviews for a tour
    public List<Review> getReviewsByTourId(Long tourId) {
        return reviewRepository.findByTourId(tourId);
    }

    // Get reviews by user
    public List<Review> getReviewsByUserId(Long userId) {
        return reviewRepository.findByUserId(userId);
    }

    // Get average rating for tour
    public Double getAverageRating(Long tourId) {
        Double avgRating = reviewRepository.getAverageRatingByTourId(tourId);
        return avgRating != null ? Math.round(avgRating * 10.0) / 10.0 : 0.0;
    }

    // Get review count for tour
    public Long getReviewCount(Long tourId) {
        return reviewRepository.countByTourId(tourId);
    }

    // Update review
    @Transactional
    public Review updateReview(Long reviewId, Long userId, Integer rating, String comment) {
        Review review = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new ResourceNotFoundException("Review not found with id: " + reviewId));

        // Check if user owns this review
        if (!review.getUser().getId().equals(userId)) {
            throw new BadRequestException("You can only edit your own reviews!");
        }

        review.setRating(rating);
        review.setComment(comment);

        return reviewRepository.save(review);
    }

    // Delete review
    @Transactional
    public void deleteReview(Long reviewId, Long userId, boolean isAdmin) {
        Review review = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new ResourceNotFoundException("Review not found with id: " + reviewId));

        // Admin can delete any review, users can only delete their own
        if (!isAdmin && !review.getUser().getId().equals(userId)) {
            throw new BadRequestException("You can only delete your own reviews!");
        }

        reviewRepository.deleteById(reviewId);
    }
}